src/data_api/events.se

```diff
@@ -1,346 +1,420 @@
+    self.events[eventId].allOutcomesTokenSupply[branch] = safeAdd(startingAllOutcomesTokenSupply, attotoken)
+
+# if another event caused a fork after users had bought outcome tokens on this branch, allow them to be traded back in for REP at face value on this branch
+def privateTryRedeemTokensForNonForkingEventOnForkedBranch(eventId, branch, outcome, tokenHolderAddress):
+    privateOnly()
+    forkingEventId = self.controller.lookup('branches').getForkingEvent(branch)
+    if (forkingEventId == 0):
+        return(0)
+    if (forkingEventId == eventId):
+        return(0)
+    tokensToRedeem = self.events[eventId].tokens[branch][outcome][tokenHolderAddress]
+    self.events[eventId].tokens[branch][outcome][tokenHolderAddress] = 0
+    # self.controller.lookup('reputationToken').transfer(branch, tokenHolderAddress, tokensToRedeem)
     return(1)
 
-def getRejected(event):
-    return(self.Events[event].rejected)
-
-def getRejectedPeriod(event):
-    return(self.Events[event].rejectedPeriod)
-
-def setExtraBondPoster(event, poster):
-    self.controller.assertIsWhitelisted(msg.sender)
-    self.Events[event].extraBondPoster = poster
+# if this event is the cause of a fork, force tokens to be redeemed on the branch associated with the token's outcome
+def privateTryRedeemTokensOnForkingEvent(eventId, branch, outcome, tokenHolderAddress):
+    privateOnly()
+    # FIXME: need to redeem fees, but the current implementation just redeems tokens on outcome branch. CONSIDER: change this to only allow token redemption after event is finalized, then we can also distribute fees on correct outcome branch
+    forkingEventId = self.controller.lookup('branches').getForkingEvent(branch)
+    if (forkingEventId == 0):
+        return(0)
+    if (forkingEventId != eventId):
+        return(0)
+    parentBranchBits = sha3(branch) & 0xffffffffffffffff000000000000000000000000000000000000000000000000
+    eventBits = sha3(eventId) & 0x0000000000000000ffffffffffffffff00000000000000000000000000000000
+    outcomeBits = sha3(outcome) & 0x00000000000000000000000000000000ffffffffffffffffffffffffffffffff
+    outcomeBranch = parentBranchBits | eventBits | outcomeBits
+    attotokens = self.events[eventId].tokens[branch][outcome][tokenHolderAddress]
+    # FIXME something seems fishy here
+    self.controller.lookup('reputationToken').migrate(branch, outcomeBranch, attotokens)
+    # self.controller.lookup('reputationToken').transfer(outcomeBranch, tokenHolderAddress, attotokens)
     return(1)
 
-def getExtraBondPoster(event):
-    return(self.Events[event].extraBondPoster)
-
-def setController(newController: address):
-    if(msg.sender != self.controller):
+def privateRedeemCorrectOutcomeTokens(eventId, branch, outcome, tokenHolderAddress):
+    privateOnly()
+    # FIXME: need to deal with fees here as well, remaining_fees_in_period / remaining_correct_outcome_tokens_in_period
```

Comment:
I also think it's not considering failed bonds [ones where the final outcome was no different than the outcome when posted] in the rep redistribution here as well